cirq-core/cirq/ops/boolean_hamiltonian.py

```diff
@@ -0,0 +1,372 @@
+    example, this object could represent the polynomial 0.5*I - 0.5*Z_1*Z_2.
+
+    Args:
+        boolean_expr: A Sympy expression containing symbols and Boolean operations
+        qubit_map: map of string (boolean variable name) to qubit.
+
+    Return:
+        The HamiltonianPolynomial that represents the Boolean expression.
+    """
+    if isinstance(boolean_expr, Symbol):
+        # Table 1 of [1], entry for 'x' is '1/2.I - 1/2.Z'
+        return 0.5 * _hamiltonian_I() - 0.5 * _hamiltonian_Z(qubit_map[boolean_expr.name])
+
+    if isinstance(boolean_expr, (And, Not, Or, Xor)):
+        sub_hamiltonians = [
+            _build_hamiltonian_from_boolean(sub_boolean_expr, qubit_map)
+            for sub_boolean_expr in boolean_expr.args
+        ]
+        # We apply the equalities of theorem 1 of [1].
+        if isinstance(boolean_expr, And):
+            hamiltonian = _hamiltonian_I()
+            for sub_hamiltonian in sub_hamiltonians:
+                hamiltonian = hamiltonian * sub_hamiltonian
+        elif isinstance(boolean_expr, Not):
+            assert len(sub_hamiltonians) == 1
+            hamiltonian = _hamiltonian_I() - sub_hamiltonians[0]
+        elif isinstance(boolean_expr, Or):
+            hamiltonian = _hamiltonian_O()
+            for sub_hamiltonian in sub_hamiltonians:
+                hamiltonian = hamiltonian + sub_hamiltonian - hamiltonian * sub_hamiltonian
+        elif isinstance(boolean_expr, Xor):
+            hamiltonian = _hamiltonian_O()
+            for sub_hamiltonian in sub_hamiltonians:
+                hamiltonian = hamiltonian + sub_hamiltonian - 2.0 * hamiltonian * sub_hamiltonian
+        return hamiltonian
+
+    raise ValueError(f'Unsupported type: {type(boolean_expr)}')
+
+
+def _gray_code_comparator(k1: Tuple[int, ...], k2: Tuple[int, ...], flip: bool = False) -> int:
+    """Compares two Gray-encoded binary numbers.
+
+    Args:
+        k1: A tuple of ints, representing the bits that are one. For example, 6 would be (1, 2).
+        k2: The second number, represented similarly as k1.
+        flip: Whether to flip the comparison.
+
+    Returns:
+        -1 if k1 < k2 (or +1 if flip is true)
+        0 if k1 == k2
+        +1 if k1 > k2 (or -1 if flip is true)
+    """
+    max_1 = k1[-1] if k1 else -1
+    max_2 = k2[-1] if k2 else -1
+    if max_1 != max_2:
+        return -1 if (max_1 < max_2) ^ flip else 1
+    if max_1 == -1:
+        return 0
+    return _gray_code_comparator(k1[0:-1], k2[0:-1], not flip)
+
+
+def _simplify_cnots_pairs(
+    cnots: List[Tuple[int, int]], flip_control_and_target: bool
+) -> Tuple[bool, List[Tuple[int, int]]]:
+    """Simplifies CNOT pairs according to equations 9 and 10 of [4].Simplifies
+
+    CNOT(i, j) @ CNOT(k, j) = CNOT(k, j) @ CNOT(i, j)
+    ───@───────       ───────@───
+       │                     │
+    ───X───X───   =   ───X───X───
+           │             │
+    ───────@───       ───@───────
+
```

Comment:
This is confusing: how come this is called `simplify_cnot_pairs`? This is a commutation relation, not a simplification. Can you clarify?

Reply:
Changed name to `_simplify_commuting_cnots`